core/ic/src/canisters/gateway.rs

```diff
@@ -146,6 +146,24 @@ fn set_chain(
     Ok(true)
 }
 
+#[update(guard = "is_authorized")]
+#[candid_method(update, rename = "add_urls")]
+fn add_urls(
+    urls: Vec<String>
+) -> Result<bool, String> {
+    // set chain config
+    let rpc_urls = CHAINS.with(|c| {
+        let mut c = c.borrow_mut();
+        c.add_urls(urls);
+        c.rpc_urls()
+    });
+    STATE_MACHINE.with(|s| {
+        let mut state_machine = s.borrow_mut();
+        state_machine.set_rpc_urls(rpc_urls);
```

Comment:
The `urls` in `StateMachine` are randomly selected one that need to call in each specific round. In each round of `Init`, the `urls` field will be set. Thus I think no need to be set in this api.